src/Stratis.Features.FederatedPeg/TargetChain/InputConsolidator.cs

```
@@ -0,0 +1,250 @@
+
+                this.PartialTransaction = SigningUtils.CombineSignatures(builder, this.PartialTransaction, new []{incomingPartialTransaction});
+
+                if (oldTransaction.GetHash() == this.PartialTransaction.GetHash())
+                {
+                    // Signing didn't work if the hash is still the same
+                    this.logger.LogDebug("Signing failed.");
+                    return ConsolidationSignatureResult.Failed();
+                }
+
+                this.logger.LogDebug("Successfully signed transaction.");
+
+                // NOTE: We don't need to reserve the transaction. The wallet will be at a standstill whilst this is happening.
+
+                // If it is FullySigned, broadcast.
+                if (this.walletManager.ValidateConsolidatingTransaction(this.PartialTransaction, true))
+                {
+                    this.logger.LogDebug("Consolidation transaction is fully signed. Broadcasting {0}", this.PartialTransaction.GetHash());
+                    this.broadcasterManager.BroadcastTransactionAsync(this.PartialTransaction);
+                    this.fullySigned = true;
+                }
+
+                return ConsolidationSignatureResult.Succeeded(this.PartialTransaction);
+            }
+        }
+
+        /// <summary>
+        /// Build a consolidating transaction.
+        /// </summary>
+        public Transaction BuildConsolidatingTransaction()
+        {
+            try
+            {
+                // TODO: Confifm that we can remove the ordering below.
+
+                List<UnspentOutputReference> unspentOutputs = this.walletManager.GetSpendableTransactionsInWallet(WithdrawalTransactionBuilder.MinConfirmations).ToList();
+
+                if (unspentOutputs.Count < WithdrawalTransactionBuilder.MaxInputs)
+                    throw new Exception("We shouldn't be consolidating transactions if we have less than 50 UTXOs to spend.");
```

Should we throw an exception like this here? Why not just return null?

---

Happy to change, see comment on try catch

---

Yeah I reckon change this so that we log the values (`unspentOutputs.Count` and `WithdrawalTransactionBuilder.MaxInputs`) and then return null.